### PR TITLE
K8SPSMDB-515: Allow setting TLS mode for mongod

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPSMDB-515](https://jira.percona.com/browse/K8SPSMDB-515) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| operator-self-healing | passed |
| smart-update | passed |
| operator-self-healing-chaos | passed |
| version-service | failed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| users | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/ab1d88b0743cb58955c959ff966449c4f31e80eb
image: `perconalab/percona-server-mongodb-operator:PR-730-ab1d88b0`

File: deploy/cr.yaml [hunk 1]
```diff
@@ -280,6 +280,8 @@ spec:
     net:
       port: 27017
       hostPort: 0
+#     tls:
+#       mode: preferSSL
```
Comment (spron-in) [inline deploy/cr.yaml:284]: We want to deprecate the whole spec.mongod section.
Also preferSSL is not only for mongod.
What about changing this flag in the custom configuration?

**Review by spron-in — CHANGES_REQUESTED** [review]

File: pkg/apis/psmdb/v1/psmdb_defaults.go [hunk 2]
```diff
@@ -70,6 +71,12 @@ func (cr *PerconaServerMongoDB) CheckNSetDefaults(platform version.Platform, log
 	if cr.Spec.Mongod.Net.Port == 0 {
 		cr.Spec.Mongod.Net.Port = defaultMongodPort
 	}
+	if cr.Spec.Mongod.Net.TLS == nil {
+		cr.Spec.Mongod.Net.TLS = &MongodSpecNetTLS{}
```
Comment (qJkee) [inline pkg/apis/psmdb/v1/psmdb_defaults.go:75]: we can auto fill mode there. Next if don't needed

File: deploy/cr.yaml [hunk 3]
```diff
@@ -5,11 +5,12 @@ metadata:
 #  finalizers:
 #    - delete-psmdb-pvc
 spec:
+  initImage: perconalab/percona-server-mongodb-operator:k8spsmdb-515
 #  platform: openshift
 #  clusterServiceDNSSuffix: svc.cluster.local
 #  pause: true
   crVersion: 1.9.0
-  image: percona/percona-server-mongodb:4.4.6-8
+  image: perconalab/percona-server-mongodb-operator:k8spsmdb-515-mongod4.4
```
Comment (nonemax) [inline deploy/cr.yaml:13]: And this to

File: deploy/cr.yaml [hunk 4]
```diff
@@ -5,11 +5,12 @@ metadata:
 #  finalizers:
 #    - delete-psmdb-pvc
 spec:
+  initImage: perconalab/percona-server-mongodb-operator:k8spsmdb-515
```
Comment (nonemax) [inline deploy/cr.yaml:8]: I think you forgot to remove this image

**Review by nonemax — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | failed |
| upgrade-consistency | failed |
| one-pod | failed |
| demand-backup | failed |
| self-healing | passed |
| limits | failed |
| scaling | passed |
| scheduled-backup | failed |
| security-context | failed |
| self-healing-chaos | passed |
| demand-backup-sharded | failed |
| rs-shard-migration | failed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| monitoring-2-0 | failed |
| arbiter | failed |
| service-per-pod | failed |
| upgrade | failed |
| liveness | failed |
| upgrade-sharded | failed |
| smart-update | failed |
| version-service | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/e786030a59bece08b5b1dd1b0e4eb45cbb128ac8
image: `perconalab/percona-server-mongodb-operator:PR-730-e786030a`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| security-context | passed |
| self-healing-chaos | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade | failed |
| upgrade-sharded | failed |
| pitr | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| pitr-sharded | failed |
| smart-update | passed |
| version-service | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/839478d1f04563a06d5acd4e6f04f405e8eebb35
image: `perconalab/percona-server-mongodb-operator:PR-730-839478d1`

**Review by delgod — CHANGES_REQUESTED** [review]

File: build/ps-entry.sh [hunk 5]
```diff
@@ -401,6 +401,11 @@ if [[ "$originalArgOne" == mongo* ]]; then
 	if [ "$MONGODB_VERSION" == 'v4.2' ] || [ "$MONGODB_VERSION" == 'v4.4' ]; then
 		_mongod_hack_rename_arg_save_val --sslMode --tlsMode "${mongodHackedArgs[@]}"
 
+                if _parse_config "$@"; then
+                    tlsMode=$(jq -r '.net.tls.mode // "preferSSL"' "$jsonConfigFile")
+                    _mongod_hack_ensure_arg_val --tlsMode "$tlsMode" "${mongodHackedArgs[@]}"
```
[the remark this answers is not in the document]
Comment (egegunes) [reply build/ps-entry.sh:406]: no, we didn't change it. if user doesn't pass the tls mode with custom config, we use `preferTLS` as default: https://github.com/percona/percona-server-mongodb-operator/blob/K8SPSMDB-515/build/ps-entry.sh#L425-L430

[the remark this answers is not in the document]
Comment (delgod) [reply build/ps-entry.sh:406]: lets have a call

**Review by qJkee — DISMISSED** [review]

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| one-pod | failed |
| self-healing | passed |
| demand-backup | failed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| operator-self-healing-chaos | passed |
| demand-backup-sharded | failed |
| upgrade | failed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/0c3b42b21a101bb58beac25b6610edd0d858dfa9
image: `perconalab/percona-server-mongodb-operator:PR-730-0c3b42b2`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | failed |
| upgrade-consistency | passed |
| demand-backup | failed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| operator-self-healing-chaos | passed |
| demand-backup-sharded | failed |
| monitoring-2-0 | passed |
| upgrade | failed |
| rs-shard-migration | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/3cd2626366e29556e8b7245f08169ac3a6fef52d
image: `perconalab/percona-server-mongodb-operator:PR-730-3cd26263`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | failed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | passed |
| rs-shard-migration | failed |
| operator-self-healing | passed |
| liveness | passed |
| smart-update | passed |
| operator-self-healing-chaos | passed |
| scheduled-backup | failed |
| version-service | failed |
| demand-backup-sharded | failed |
| users | failed |
| data-sharded | passed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/338dba79230b7dadbe93c01890f9d55daf3243e2
image: `perconalab/percona-server-mongodb-operator:PR-730-338dba79`

File: pkg/controller/perconaservermongodb/mgo.go [hunk 6]
```diff
@@ -318,11 +321,29 @@ func (r *ReconcilePerconaServerMongoDB) handleReplsetInit(m *api.PerconaServerMo
...
+		mongoVer := v.Must(v.NewVersion(strings.TrimSpace(outb.String())))
+
+		mongoCmd := "mongo"
+		if !m.Spec.UnsafeConf {
+			// --tls* options is introduced in 4.2, we need to use --ssl* for older versions
+			if mongoVer.Compare(sslDeprecatedVersion) < 0 {
+				mongoCmd += " --ssl --sslPEMKeyFile /tmp/tls.pem --sslAllowInvalidCertificates --sslCAFile /etc/mongodb-ssl/ca.crt"
+			} else {
+				mongoCmd += " --tls --tlsCertificateKeyFile /tmp/tls.pem --tlsAllowInvalidCertificates --tlsCAFile /etc/mongodb-ssl/ca.crt"
+			}
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/mgo.go:339]: but these transformations are done inside the entrypoint. see - https://github.com/percona/percona-server-mongodb-operator/blob/main/build/ps-entry.sh#L420

I am not sure if we should duplicate logic here?
```suggestion
			mongoCmd += " --ssl --sslPEMKeyFile /tmp/tls.pem --sslAllowInvalidCertificates --sslCAFile /etc/mongodb-ssl/ca.crt"
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | failed |
| limits | passed |
| data-sharded | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | failed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/c205289b7ebfcb8befe1a8228e405d43c4ed1be0
image: `perconalab/percona-server-mongodb-operator:PR-730-c205289b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | failed |
| init-deploy | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| limits | passed |
| arbiter | passed |
| demand-backup | failed |
| scaling | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| scheduled-backup | failed |
| users | failed |
| demand-backup-sharded | failed |
| data-sharded | passed |
| non-voting | failed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/a6b37284f3c64b645d592b66c5934a919cb99b1f
image: `perconalab/percona-server-mongodb-operator:PR-730-a6b37284`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | failed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | failed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| scheduled-backup | failed |
| users | failed |
| data-sharded | passed |
| demand-backup-sharded | failed |
| non-voting | failed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/730/commits/77f0f97894b8702e3835da26477a2fe955b871bd
image: `perconalab/percona-server-mongodb-operator:PR-730-77f0f978`

**Comment by egegunes** [comment]
Closing this until I'm available to give another shot.

**Pull request closed without merging**